### PR TITLE
Replaced SoftReference with WeakReference for better performance

### PR DESCRIPTION
**Pull request by orionll**
Benchmark:

```java
import com.google.common.base.Stopwatch;
import fj.data.Stream;

public class Main {
    public static void main(String[] args) {
        int to = 3000000;
        Stream<Integer> range = Stream.range(1, 3000000);

        Stopwatch sw = Stopwatch.createStarted();
        Integer sum = range.foldLeft(x -> y -> x + y, 0);
        System.out.println("First run: " + sw.stop());

        sw.start();
        sum = range.foldLeft(x -> y -> x + y, 0);
        System.out.println("Second run: " + sw.stop());
    }
}
```

The result with SoftReference:
```
First run: 13,31 s
Second run: 32,68 s
```

WeakReference:
```
First run: 219,9 ms
Second run: 388,6 ms
```

SoftReference is about 70 times slower.

I also performed a profiling session with JVisualVM. With SoftReference there are 125 collections total and garbage collection took 23 seconds! With WeakReference - 29 collections and 25 milliseconds.

### CODE REVIEW

**Comment by jbgi** [comment]
but with WeakReference there almost not point to memoize at all (the result can be garaged really fast). Might as well remove the call memo() from the stream implementation.
Another benchmark taking advantage of stream memoization would be useful.

**Comment by orionll** [comment]
We use a cache of parsed expressions in production (`com.google.common.cache.Cache<String, ASTNode>`). We tested different implementations of cache (WeakReference-based, SoftReference-based) and also no caching at all (each expression is parsed every time). The SoftReference-based cache showed terrible performance, even worse than no caching. The WeakReference-based cache showed the best performance, and we are using it in production now.

**Comment by jbgi** [comment]
I think there is no general rule, it depends of how expensive it is to re-compute cached values and how frequently you need them. Maybe we need the three memo implementation to let the user chose the most appropriate one:
 - hardMemo() (basically com.google.common.base.Suppliers.MemoizingSupplier) 
 - softMemo() (current implementation) 
 - weakMemo() (your patch) 
It is totally possible that for most use case of streams weakMemo give the best results.

**Comment by jbgi** [comment]
What I meant is that we should not change the current ```memo()``` behavior (which would be an alias for ```softMemo()```) and maybe instead change Stream implementation to use a new ```weakMemo()``` method on P1.

**Comment by orionll** [comment]
+1 for adding three distinct methods and removing memo().

**Comment by orionll** [comment]
So, I think there should be three methods:
* memo() - based on strong references
* weakMemo() - based on weak references
* softMemo() - based on soft references

See my changes.

**Comment by jbgi** [comment]
@mperry can we change memo() memory behaviour without a full deprecation cycle?

**Comment by techtangents** [comment]
How about we keep memo(), but add an overload for it that passes in the 
memoization strategy.

e.g.
memo(Memo.hardReference)
memo(Memo.softReference)
memo(Memo.weakReference)

then memo() would be implemented as:

memo(Memo.weakReference)

or whichever behavior you want as default.

**Comment by orionll** [comment]
I don't see any benefist of it

File: quickcheck/src/main/java/fj/test/Gen.java [hunk 1]
```diff
@@ -337,15 +337,7 @@ public A f(final Rand r) {
...
   public static <A> Gen<List<A>> sequence(final List<Gen<A>> gs) {
-    return gs.foldRight(new F<Gen<A>, F<Gen<List<A>>, Gen<List<A>>>>() {
-      public F<Gen<List<A>>, Gen<List<A>>> f(final Gen<A> ga) {
-        return new F<Gen<List<A>>, Gen<List<A>>>() {
-          public Gen<List<A>> f(final Gen<List<A>> gas) {
-            return ga.bind(gas, List.<A>cons());
-          }
-        };
-      }
-    }, value(List.<A>nil()));
```
Comment (orionll) [inline quickcheck/src/main/java/fj/test/Gen.java]: This caused StackOverflow

File: core/src/main/java/fj/data/List.java [hunk 2]
```diff
@@ -625,18 +624,18 @@ public final void foreachDoEffect(final Effect1<A> f) {
    * @return A new list that has appended the given list.
    */
   public final List<A> append(final List<A> as) {
-    return fromList(this).append(as).toList();
+    return Buffer.fromList(this).prependToList(as);
```
Comment (orionll) [inline core/src/main/java/fj/data/List.java]: List.append() was O(n+m). I made it O(n)

**Comment by mperry** [comment]
I agree with @techtangents, I think we should provide a default memo implementation and then other options to choose from a hard, soft and weak implementation.  Whether this is separate hard, soft and weak methods or a parameter to memo does not matter too much.

**Comment by mperry** [comment]
@jbgi we can change the implementation of `memo` memory implementation without deprecation.  We don't provide guarantees on the implementation, so changing the implementation when the type signature remains the same I think is fine.

**Comment by mperry** [comment]
I think this issue is complete enough to merge.  I created #153 to handle hard, soft, weak and default memo implementations.

**Pull request merged**